### PR TITLE
[REVIEW] Generalize `describe` & `quantile` apis

### PR DESCRIPTION
**Pull request by galipremsagar**
This fixes: #5136, #5138 
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: dask/dataframe/methods.py [hunk 1]
```diff
@@ -152,8 +152,11 @@ def describe_aggregate(values):
 def describe_numeric_aggregate(stats, name=None, is_timedelta_col=False):
     assert len(stats) == 6
     count, mean, std, min, q, max = stats
+    import sys
 
-    typ = pd.DataFrame if isinstance(count, pd.Series) else pd.Series
+    package_name = count.__class__.__module__.split(".")[0]
+    DD = sys.modules[package_name]
+    typ = DD.DataFrame if is_series_like(count) else DD.Series
```
Comment (mrocklin) [inline dask/dataframe/methods.py]: Perhaps the following instead? 

```python
typ = type(count.to_frame()) if is_series_like(count) else type(count)
```

This doesn't dive as deeply into internals

File: dask/dataframe/methods.py [hunk 2]
```diff
@@ -162,15 +165,18 @@ def describe_numeric_aggregate(stats, name=None, is_timedelta_col=False):
         max = pd.to_timedelta(max)
         q = q.apply(lambda x: pd.to_timedelta(x))
 
-    part1 = typ([count, mean, std, min], index=["count", "mean", "std", "min"])
+    part1 = typ(
+        [count.astype("float64"), mean, std, min], index=["count", "mean", "std", "min"]
```
Comment (mrocklin) [inline dask/dataframe/methods.py]: Why this change?

File: dask/dataframe/methods.py [hunk 3]
```diff
@@ -162,15 +165,18 @@ def describe_numeric_aggregate(stats, name=None, is_timedelta_col=False):
...
+    )
+
     q.index = ["{0:g}%".format(l * 100) for l in q.index.tolist()]
-    if isinstance(q, pd.Series) and typ == pd.DataFrame:
+    if is_series_like(q) and is_dataframe_like(typ):
         q = q.to_frame()
     part3 = typ([max], index=["max"])
 
-    result = pd.concat([part1, q, part3], **concat_kwargs)
+    result = DD.concat([part1, q, part3], **concat_kwargs)
```
Comment (mrocklin) [inline dask/dataframe/methods.py]: Perhaps use the concat function defined in this module instead

File: dask/dataframe/methods.py (hunk 2 above)
Comment (TomAugspurger) [reply dask/dataframe/methods.py]: Note that pandas changed behavior here in 0.25.0: https://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.25.0.html#dataframe-describe-on-an-empty-categorical-object-column-will-return-top-and-freq / https://github.com/pandas-dev/pandas/pull/26474. We deliberately have object-type, so that the `NaN` for `top` and `freq` don't convert the integer `count` and `unique` to float.

Comment (galipremsagar) [reply dask/dataframe/methods.py]: @mrocklin The count type is an odd one out - int, while rest all are floats, because of this a transpose operation in cudf.DataFrame will start failing. While there is no such restriction on pandas side though. It's a cudf limitation which motivated me to make this typecast in dask.

File: dask/dataframe/methods.py (hunk 1 above)
Comment (galipremsagar) [reply dask/dataframe/methods.py]: Changed this part

File: dask/dataframe/methods.py (hunk 3 above)
Comment (galipremsagar) [reply dask/dataframe/methods.py]: Using the concat function in this module, but to preserve `concat_kwargs` added it to the `concat` method too. Adding support for cudf backend too as a stale function param.

File: dask/dataframe/methods.py [hunk 4]
```diff
@@ -333,7 +330,7 @@ def pivot_count(df, index, columns, values):
 concat_dispatch = Dispatch("concat")
 
 
-def concat(dfs, axis=0, join="outer", uniform=False, filter_warning=True):
+def concat(dfs, axis=0, join="outer", uniform=False, filter_warning=True, **concat_kwargs):
```
Comment (mrocklin) [inline dask/dataframe/methods.py]: Lets just use `**kwargs` here and below.  I think that we don't need to specify the `concat_` bit in the name.

File: dask/dataframe/methods.py (hunk 2 above)
Comment (mrocklin) [reply dask/dataframe/methods.py]: > It's a cudf limitation which motivated me to make this typecast in dask.

How hard would it be to fix this within cudf?

In general I think that it makes sense for counts to be integers rather than floats.  I would prefer to see cudf gain functionality rather than see dask.dataframe lose functionality as the two grow together if possible.

@TomAugspurger I'm not sure I understand your comment.

Comment (TomAugspurger) [reply dask/dataframe/methods.py]: In pandas 0.24, this was the output

```python
In [4]: pd.DataFrame(columns=['A']).describe()
Out[4]:
        A
count   0
unique  0

In [5]: pd.DataFrame(columns=['A']).describe().dtypes
Out[5]:
A    int64
dtype: object
```

In 0.25, it's

```python
In [4]: pd.DataFrame(columns=['A']).describe()
Out[4]:
          A
count     0
unique    0
top     NaN
freq    NaN

In [5]: pd.DataFrame(columns=['A']).describe().dtypes
Out[5]:
A    object
dtype: object
```

If Out[5] weren't object dtype, the pretense of NaNs would convert the output to float64 dtype.

File: dask/dataframe/methods.py (hunk 4 above)
Comment (galipremsagar) [reply dask/dataframe/methods.py]: Done

File: dask/dataframe/core.py [hunk 5]
```diff
@@ -4949,16 +4949,21 @@ def quantile(df, q, method="default"):
     else:
         internal_method = method
 
+    import sys
+
+    package_name = df.__class__.__module__.split(".")[0]
+    DD = sys.modules[package_name]
```
Comment (galipremsagar) [inline dask/dataframe/core.py]: Had to fetch package as I didn't find a way to construct a `Series` object with a `DataFrame` in hand.

Comment (mrocklin) [reply dask/dataframe/core.py]: I would prefer that we not fetch packages like this.  You can get a series from a dataframe by accessing one of the columns:

```python
s = df[df.columns[0]]
```

File: dask/dataframe/methods.py (hunk 2 above)
Comment (galipremsagar) [reply dask/dataframe/methods.py]: > > It's a cudf limitation which motivated me to make this typecast in dask.
> 
> How hard would it be to fix this within cudf?

It would be quite a huge change(not an easy one) which wouldn't be happening anytime soon as it involves re-work in cpp layer and still there isn't a solution figured out to do that change at cpp layer.

Comment (mrocklin) [reply dask/dataframe/methods.py]: OK, we might have to get creative then.  Generally I don't think that Dask should degrade performance in the common case (Pandas) in order to work around a cudf limitation.  My guess is that there is still a nice way around this problem.  I don't know what it is of course, but my guess is that it's doable :)

Comment (galipremsagar) [reply dask/dataframe/methods.py]: Two options that come to my mind:
1. Encapsulating this line in a try/catch block and incase of an error we retry by typecasting.
2. Have a check of on `typ`, if it belongs to cudf module and then typecaste?

Comment (mrocklin) [reply dask/dataframe/methods.py]: Sorry, but generally we don't special case libraries like cudf within dask.dataframe today.  Explicit type checks aren't an option.

I think that there must be a way to turn a few cudf Series objects of different types into a cudf DataFrame in a way that is also Pandas compliant.  If there isn't then this seems like something that we should fix on the cudf side.  Is there an open issue about this somewhere?  If not, can we make one?

Putting on my Dask maintainer hat, this operation of constructing a Series from a list of scalars or a DataFrame from a list of Series objects seems fairly routine.  I'm somewhat against inserting special case code here.

Comment (mrocklin) [reply dask/dataframe/methods.py]: cc @kkraus14 , any suggestions?

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -4951,14 +4951,19 @@ def quantile(df, q, method="default"):
 
     # currently, only Series has quantile method
     if isinstance(df, Index):
-        meta = pd.Series(df._meta_nonempty).quantile(q)
+        series_typ = type(df._meta.to_series())
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Can this just be `df._meta.to_series().quantile()`? I don't think we need the `series_type()` call, since IIUC that's doing `Series(series)`

File: dask/dataframe/core.py [hunk 7]
```diff
@@ -4951,14 +4951,19 @@ def quantile(df, q, method="default"):
...
+        meta = series_typ(df._meta_nonempty).quantile(q)
     else:
+        if is_series_like(df._meta):
+            series_typ = type(df._meta)
+        else:
+            series_typ = type(df._meta[df._meta.columns[0]])
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: This will fail for duplicate columns.

```python
df = pd.DataFrame(columns=['a', 'a'])
ddf = dd.from_pandas(df, 2)
```

File: dask/dataframe/methods.py (hunk 2 above)
Comment (TomAugspurger) [reply dask/dataframe/methods.py]: Do we have a solution to this?

Comment (mrocklin) [reply dask/dataframe/methods.py]: I think that we're still waiting on @kkraus14 .  I'll poke him through side channels.

**Comment by kkraus14** [comment]
Sorry playing catch up on this, will respond in a bit.

File: dask/dataframe/methods.py (hunk 2 above)
Comment (kkraus14) [reply dask/dataframe/methods.py]: @mrocklin @TomAugspurger We can't exactly match the behavior of Pandas from cuDF here due to the use of object types.

There's potentially a solution to have cuDF spit out `float64` columns, but the challenge is `count` actually outputs an `int64` type in cuDF. It then tries to transpose a combination of `int64` and `float64` dtypes which is where it throws an error as opposed to magic typecasting.

Comment (TomAugspurger) [reply dask/dataframe/methods.py]: Does cuDF support a nullable integer dtype? If so, I suspect that expecting that would be OK.

Comment (kkraus14) [reply dask/dataframe/methods.py]: Yes it does, but the problem is say you have a `datetime` or `string` Series that you're asking to describe, `top` would be the dtype of the Series, whereas `count`, `unique`, etc. would still be numeric types.

**Comment by TomAugspurger** [comment]
@galipremsagar is it clear what needs to be done here? Can you summarize the remaining sticking points?

**Comment by galipremsagar** [comment]
@TomAugspurger, Yes. Will be fixing the following two review comments:
1. https://github.com/dask/dask/pull/5137#discussion_r312037263
2. https://github.com/dask/dask/pull/5137#discussion_r312037774

But would require inputs from @kkraus14 or @mrocklin on how to proceed with this inconsistency between cudf and pandas:
https://github.com/dask/dask/pull/5137#discussion_r312038709

File: dask/dataframe/core.py (hunk 7 above)
Comment (galipremsagar) [reply dask/dataframe/core.py]:
```python
>>> df = pd.DataFrame(columns=['a', 'a'])
>>> ddf = dd.from_pandas(df, 2)
>>> ddf._meta
Empty DataFrame
Columns: [a, a]
Index: []
>>> ddf._meta.columns[0]
'a'
>>> ddf._meta[ddf._meta.columns[0]]
Empty DataFrame
Columns: [a, a]
Index: []
>>> type(ddf._meta[ddf._meta.columns[0]])
<class 'pandas.core.frame.DataFrame'>
>>> ddf.quantile()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gali/Documents/GitHub/dask/dask/dataframe/core.py", line 1876, in quantile
    meta = self._meta.quantile(q, axis=axis)
  File "/Users/gali/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py", line 7778, in quantile
    transposed=is_transposed)
  File "/Users/gali/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 483, in quantile
    values = _concat._concat_compat([b.values for b in blocks])
  File "/Users/gali/anaconda3/lib/python3.7/site-packages/pandas/core/dtypes/concat.py", line 170, in _concat_compat
    return np.concatenate(to_concat, axis=axis)
ValueError: need at least one array to concatenate
>>> df.quantile()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gali/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py", line 7778, in quantile
    transposed=is_transposed)
  File "/Users/gali/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 483, in quantile
    values = _concat._concat_compat([b.values for b in blocks])
  File "/Users/gali/anaconda3/lib/python3.7/site-packages/pandas/core/dtypes/concat.py", line 170, in _concat_compat
    return np.concatenate(to_concat, axis=axis)
ValueError: need at least one array to concatenate
```

@TomAugspurger The only motive of this is to extract the `DataFrame` type, so I suppose even if it has duplicate columns this will work as desired.

**Comment by jrbourbeau** [comment]
Thanks for the updates @galipremsagar! @TomAugspurger could you give this another look when you get a moment (post conference traveling)?

**Comment by beckernick** [comment]
@galipremsagar do you have the bandwidth to resolve conflicts in this PR?

**Comment by galipremsagar** [comment]
@beckernick Resolved conflicts and this is ready for review.

cc: @TomAugspurger , @jrbourbeau

File: dask/dataframe/core.py [hunk 8]
```diff
@@ -5246,14 +5246,19 @@ def quantile(df, q, method="default"):
 
     # currently, only Series has quantile method
     if isinstance(df, Index):
-        meta = pd.Series(df._meta_nonempty).quantile(q)
+        series_typ = type(df._meta.to_series())
```
Comment (TomAugspurger) [inline dask/dataframe/core.py:5249]:
```suggestion
        series_typ = df._meta.to_series()._constructor
```

I think cudf implements this now.

File: dask/dataframe/core.py [hunk 9]
```diff
@@ -5246,14 +5246,19 @@ def quantile(df, q, method="default"):
...
+        meta = df._meta_nonempty.to_series().quantile(q)
     else:
+        if is_series_like(df._meta):
+            series_typ = type(df._meta)
```
Comment (TomAugspurger) [inline dask/dataframe/core.py:5253]:
```suggestion
            series_typ = df._meta._constructor
```

File: dask/dataframe/core.py [hunk 10]
```diff
@@ -5246,14 +5246,19 @@ def quantile(df, q, method="default"):
...
+        else:
+            series_typ = type(df._meta[df._meta.columns[0]])
```
Comment (TomAugspurger) [inline dask/dataframe/core.py:5255]:
```suggestion
            series_typ = df._meta._constructor_sliced
```

File: dask/dataframe/methods.py [hunk 11]
```diff
@@ -153,7 +153,10 @@ def describe_numeric_aggregate(stats, name=None, is_timedelta_col=False):
     assert len(stats) == 6
     count, mean, std, min, q, max = stats
 
-    typ = pd.DataFrame if isinstance(count, pd.Series) else pd.Series
+    if is_series_like(count) and hasattr(count, "to_frame"):
+        typ = type(count.to_frame())
+    else:
+        typ = type(q)
```
Comment (TomAugspurger) [inline dask/dataframe/methods.py:159]: Why is the hasattr needed?

File: dask/dataframe/methods.py [hunk 12]
```diff
@@ -351,7 +355,13 @@ def pivot_count(df, index, columns, values):
...
 
 def concat(
-    dfs, axis=0, join="outer", uniform=False, filter_warning=True, ignore_index=False
+    dfs,
+    axis=0,
+    join="outer",
+    uniform=False,
+    filter_warning=True,
+    ignore_index=False,
+    **kwargs
```
Comment (TomAugspurger) [inline dask/dataframe/methods.py:364]: What's the motivation for these changes?

File: dask/dataframe/tests/test_multi.py [hunk 13]
```diff
@@ -560,24 +560,29 @@ def test_concat(join):
...
-        (ddf1, ddf3, pdf1, pdf3),
-        (ddf1.x, ddf2.x, pdf1.x, pdf2.x),
-        (ddf1.x, ddf3.z, pdf1.x, pdf3.z),
-        (ddf1.x, ddf2.x, pdf1.x, pdf2.x),
-        (ddf1.x, ddf3.z, pdf1.x, pdf3.z),
-    ]:
-        expected = pd.concat([pd1, pd2], **kwargs)
-        result = dd.concat([dd1, dd2])
-        assert_eq(result, expected)
+    with warnings.catch_warnings():
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_multi.py:563]: I don't think we want to change this. We need to ensure that users aren't seeing pandas warnings from Dask's usage of `pandas.concat`.

Comment (galipremsagar) [reply dask/dataframe/tests/test_multi.py:563]: Should we then filter warnings in `dd.concat`?

File: dask/dataframe/methods.py (hunk 12 above)
Comment (galipremsagar) [reply dask/dataframe/methods.py:364]: This was to be able to send in additional arguments like `sort=False/True` with flexibility for the caller to choose.

File: dask/dataframe/methods.py (hunk 11 above)
Comment (galipremsagar) [reply dask/dataframe/methods.py]: Actually this is redundant, removed the check.

File: dask/dataframe/tests/test_multi.py (hunk 13 above)
Comment (TomAugspurger) [reply dask/dataframe/tests/test_multi.py:563]: No, I don't think so. We need to ensure that we call `pd.concat` with the correct parameters. I think that the test as written gets the behavior that we want.

Comment (galipremsagar) [reply dask/dataframe/tests/test_multi.py]: Removed the warnings filtering in tests, upon some digging found that the param wasn't being propagated to the functions being called in the code-flow.

**Comment by TomAugspurger** [comment]
Thanks @galipremsagar!
